chore(docs): update docs and prompt to emphasize pushing of both yml files

huanjani opened the pull request:
Previously, only the `buildspec.yml` needed to be added to the source repo in order to build a pipeline. But now the buildspec checks the pipeline manifest to see which environment stages to deploy to (and only upgrades those envs), so both files need to be checked into the user's repo. This updates docs and CLI prompts to reflect that change. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

huanjani committed:
update: edit docs and prompt to emphasize pushing of both yml files

huanjani committed:
fix(docs): add colon, full word for 'repo'

dependabot[bot] committed:
chore: Bump github.com/aws/aws-sdk-go from 1.37.10 to 1.37.15 (#1951)

Bumps [github.com/aws/aws-sdk-go](https://github.com/aws/aws-sdk-go) from 1.37.10 to 1.37.15.
- [Release notes](https://github.com/aws/aws-sdk-go/releases)
- [Changelog](https://github.com/aws/aws-sdk-go/blob/master/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-go/compare/v1.37.10...v1.37.15)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore: Bump github.com/golang/mock from 1.4.4 to 1.5.0 (#1953)

Bumps [github.com/golang/mock](https://github.com/golang/mock) from 1.4.4 to 1.5.0.
- [Release notes](https://github.com/golang/mock/releases)
- [Changelog](https://github.com/golang/mock/blob/master/.goreleaser.yml)
- [Commits](https://github.com/golang/mock/compare/v1.4.4...v1.5.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

bvtujo commented on site/content/docs/concepts/pipelines.md:
```diff
@@ -93,27 +93,31 @@ stages:
 When this buildspec runs, it pulls down the version of Copilot which was used when you ran `pipeline init`, to ensure backwards compatibility.
 
-### Step 4: Creating your Pipeline
+### Step 4: Pushing New Files to your Repository
```
Can we reword this section to include the fact that we need *everything* in the copilot directory? Because we need service manifests, addons templates, pipeline and buildspec, and *especially* `.workspace`. 

edit: Efe got there first 🥇

efekarakus commented on internal/pkg/cli/pipeline_init.go:
```diff
@@ -205,13 +205,11 @@ func (o *initPipelineOpts) Execute() error {
-		fmt.Sprintf("Update the %s phase of your buildspec to unit test your services before pushing the images.", color.HighlightResource("build")),
-		"Update your pipeline manifest to add additional stages.",
-		fmt.Sprintf("Run %s to deploy your pipeline for the repository.", color.HighlightCode("copilot pipeline update")),
+		"Commit and push the generated buildspec and pipeline manifest files to the repository.",
```
Should we just say 
```suggestion
		fmt.Sprintf("Commit and push the %s directory to your remote.", color.HighlightCode("copilot")),
```

efekarakus commented on internal/pkg/cli/pipeline_init.go:
```diff
@@ -205,13 +205,11 @@ func (o *initPipelineOpts) Execute() error {
-		"Update your pipeline manifest to add additional stages.",
-		fmt.Sprintf("Run %s to deploy your pipeline for the repository.", color.HighlightCode("copilot pipeline update")),
+		"Commit and push the generated buildspec and pipeline manifest files to the repository.",
+		fmt.Sprintf("Run %s to deploy your pipeline for the specified repository and branch.", color.HighlightCode("copilot pipeline update")),
```
Maybe to be more concise?
```suggestion
		fmt.Sprintf("Run %s to create your pipeline.", color.HighlightCode("copilot pipeline update")),
```

efekarakus commented on internal/pkg/cli/pipeline_init.go:
```diff
@@ -205,13 +205,11 @@ func (o *initPipelineOpts) Execute() error {
+func (o *initPipelineOpts) RequiredActions() []string {
 	return []string{
-		"Commit and push the generated buildspec and manifest file.",
-		fmt.Sprintf("Update the %s phase of your buildspec to unit test your services before pushing the images.", color.HighlightResource("build")),
-		"Update your pipeline manifest to add additional stages.",
```
Do we have this info written anywhere? I feel like this is helpful for customers to know what these files are for

```
✔ `copilot/buildspec.yml` to build your container images and run unit tests
✔ `copilot/manifest.yml` to add additional pipeline stages
Recommended follow-up actions:
- Commit and push ...
- Run copilot pipeline update ...
```

huanjani replied:
Yes, we `	log.Infoln("The buildspec contains the commands to build and push your container images to your ECR repositories.")` when the file is created/confirmed to exist; I will add the bit about adding unit tests.

huanjani replied:
See output below. Too wordy?

efekarakus commented on site/content/docs/concepts/pipelines.md:
```diff
@@ -24,14 +24,14 @@ Want to learn more about CodePipeline? Check out their [getting started docs](ht
 
 1. Preparing the pipeline structure.
-2. Committing the generated `buildspec.yml`.
+2. Committing and pushing the generated `pipeline.yml` and `buildspec.yml`.
```
The `.workspace` file needs to also be there to pick up the app name. So anything under `copilot/`

bvtujo commented on site/content/docs/concepts/pipelines.md:
````diff
@@ -24,14 +24,14 @@ Want to learn more about CodePipeline? Check out their [getting started docs](ht
 ```bash
 $ copilot pipeline init
-$ git add copilot/buildspec.yml && git commit -m "Adding Pipeline Buildspec" && git push
+$ git add copilot/pipeline.yml copilot/buildspec.yml && git commit -m "Adding Pipeline Buildspec" && git push
````
```suggestion
$ git add copilot/pipeline.yml copilot/buildspec.yml copilot/.workspace && git commit -m "Adding Pipeline Buildspec" && git push
```

huanjani replied:
Okay to change to just `git add copilot &&`...?

bvtujo replied:
Hidden files like .workspace don't get picked up automatically by `git add copilot/*` so it needs to be explicitly added, unfortunately. 

Could do `git add copilot/* copilot/.workspace && ...`

bvtujo replied:
or `git add --all copilot` also works!

huanjani replied:
Ah, thanks. I'll change to your original suggestion bc it's the most clear for the reader.

bvtujo committed:
chore(template): add container paths as environment variables  (#1926)

Adds conditional environment variables based on volume definitions for a service or job. If there is a mounted volume `myEFSVolume` at path `/var/www`, it will be injected into the template as an environment variable. Additional volumes will be added to the JSON map. 
```yaml
- Name: COPILOT_MOUNT_POINTS
  Value: '{"myEFSVolume":"/var/www"}'
```


<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

huanjani committed:
fix: tweak wording

efekarakus committed:
chore: split manifest layout to "trigger" and "containers" sections (#1942)

To group related fields together, we updated the manifest layout to the following format:
```yaml
# Trigger for your service/task.
http: ...
on: ...

# Configuration for your containers and service.
image: ...
cpu:
memory:
count:
network:
volumes:
```

This PR also prettifies our manifest pages on the website by updating the examples with the layout above and moving them into a collapsible section:
![Screen Shot 2021-02-19 at 8 47 41 AM](https://user-images.githubusercontent.com/879348/108534599-2e715080-728f-11eb-8ffb-3d56b29e5eb2.png)


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

huanjani committed:
fix: fix formatting

huanjani commented:
```
% copilot pipeline init
1st stage: test
Which repository would you like to use for your pipeline? codecommit::us-west-2://aws-sample
✔ Pipeline manifest file for aws-sample already exists at copilot/pipeline.yml, skipping writing it.
The manifest contains configurations for your CodePipeline resources, such as your pipeline stages and build steps.
Update the file to add additional stages, change the branch to be tracked, or add test commands or manual approval actions.
✔ Buildspec file for pipeline already exists at copilot/buildspec.yml, skipping writing it.
The buildspec contains the commands to build and push your container images to your ECR repositories.
Update the build phase to unit test your services before pushing the images.

Required follow-up actions:
- Commit and push the `copilot` directory to your repository.
- Run `copilot pipeline update` to create your pipeline.
jwh@38f9d33fed75 aws-copilot-sample-service %

huanjani committed:
fix: add formatting and newlines to info

huanjani committed:
fix: change git add command in example

bvtujo committed:
chore(templates): remove incorrect use of 'and' in env-vars (#1961)

<!-- Provide summary of changes -->
The "and" intrinsic function in golang templates [evaluates both arguments](https://golang.org/pkg/text/template/#hdr-Functions). This caused a nil pointer dereference in the env vars partial template when storage objects were not provided:

```go
{{if and .Storage .Storage.MountPoints}}
```
evaluates the same as "if .Storage then .Storage.MountPoints else .Storage". 

This PR fixes that by using two conditionals. 
<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

huanjani committed:
fix(cicd): add capabilities to cfn template to enable pipeline deployment w addons (#1962)

Adds necessary CloudFormation capabilities for pipeline deployment when addons stack is present.

Fixes #1825, #1954.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

huanjani committed:
chore(cli): adapt tree to new treeprint version (#1963)

With the new version of treeprint, we are now able to specify the root of the tree. Previously, we had each stage as a branch as a workaround to avoid the extra dot and lines. This change restores the output of `pipeline status` to appear as it had, leveraging the new `NewWithRoot()`. Without this change, the stages appear as branches with extra lines.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bvtujo committed:
fix(templates): fix typo in sidecar template (#1964)

<!-- Provide summary of changes -->
![](https://emojis.slackmojis.com/emojis/images/1566825215/6248/dumpster-fire.gif?1566825215)

One line change. This is a typo in a template which otherwise should execute. Currently waiting on e2es to finish locally. 

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Lou1415926 committed:
chore(cli): prompt to select app if not under a workspace (#1958)

This pull requests changes current cli behavior for `env init`. If the user is not currently in any workspace and no application is specified using `--app` flag, it prompts the user to select from existing applications instead of returning an error. 

Resolves #1238  

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

huanjani commented:
This is what the output of `pipeline init` looked like previously (our of order!):
```
% copilot pipeline init
1st stage: test
2nd stage: quiz
Which repository would you like to use for your pipeline? codecommit::us-west-2://aws-sample
✔ Wrote the pipeline manifest for aws-sample at 'copilot/pipeline.yml'
The manifest contains configurations for your CodePipeline resources, such as your pipeline stages and build steps.
✔ Buildspec file for pipeline already exists at copilot/buildspec.yml, skipping writing it.
The buildspec contains the commands to build and push your container images to your ECR repositories.

Recommended follow-up actions:
- Commit and push the generated buildspec and manifest file.
- Update the build phase of your buildspec to unit test your services before pushing the images.
- Update your pipeline manifest to add additional stages.
- Run `copilot pipeline update` to deploy your pipeline for the repository.
%

efekarakus commented on internal/pkg/cli/pipeline_init.go:
```diff
@@ -505,7 +503,7 @@ func (o *initPipelineOpts) createPipelineManifest() error {
 	}
 	log.Successf(manifestMsgFmt, color.HighlightUserInput(o.repoName), color.HighlightResource(manifestPath))
-	log.Infoln("The manifest contains configurations for your CodePipeline resources, such as your pipeline stages and build steps.")
+	log.Infoln("The manifest contains configurations for your CodePipeline resources, such as your pipeline stages and build steps.\nUpdate the file to add additional stages, change the branch to be tracked, or add test commands or manual approval actions.")
```
```suggestion
	log.Infof(`The manifest contains configurations for your CodePipeline resources, such as your pipeline stages and build steps.
Update the file to add additional stages, change the branch to be tracked, or add test commands or manual approval actions.
`)
```

efekarakus commented on internal/pkg/cli/pipeline_init.go:
```diff
@@ -545,7 +543,7 @@ func (o *initPipelineOpts) createBuildspec() error {
 	}
 	log.Successf(buildspecMsgFmt, color.HighlightResource(buildspecPath))
-	log.Infoln("The buildspec contains the commands to build and push your container images to your ECR repositories.")
+	log.Infof("The buildspec contains the commands to build and push your container images to your ECR repositories.\nUpdate the %s phase to unit test your services before pushing the images.\n", color.HighlightResource("build"))
```
```suggestion
	log.Infof(`The buildspec contains the commands to build and push your container images to your ECR repositories.
Update the %s phase to unit test your services before pushing the images.
`, color.HighlightResource("build"))
```

efekarakus commented on site/content/docs/concepts/pipelines.md:
````diff
@@ -24,14 +24,14 @@ Want to learn more about CodePipeline? Check out their [getting started docs](ht
 ```bash
 $ copilot pipeline init
-$ git add copilot/buildspec.yml && git commit -m "Adding Pipeline Buildspec" && git push
+$ git add copilot/pipeline.yml copilot/buildspec.yml copilot/.workspace && git commit -m "Adding Pipeline Buildspec" && git push
````
```suggestion
$ git add copilot/pipeline.yml copilot/buildspec.yml copilot/.workspace && git commit -m "Adding pipeline artifacts" && git push
```

efekarakus reviewed (commented):
Looks good just some minor formatting changes

huanjani committed:
Update internal/pkg/cli/pipeline_init.go

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

huanjani committed:
Update internal/pkg/cli/pipeline_init.go

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

huanjani committed:
Update site/content/docs/concepts/pipelines.md

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

bvtujo committed:
chore(stream): add exponential backoff and jitter to Fetch (#1935)

<!-- Provide summary of changes -->
this jithub pull request adds jitter and exponential backoff to the calls to Fetch in both Cloudformation and ECS streamers when they encounter a throttling exception. 

The new interval with jitter has an average duration of 2s, with exponential backoff for throttling exceptions up to 32s. It uses uniformly distributed random numbers between 0 and 4000ms for the normal calls, and if it encounters a throttle, it will return nil and retry with a random number between 0 and 8000ms, etc. etc. 

The advantage of using a random interval with 0 minimum is that the calls from many simultaneous clients are more smoothly distributed, and exponential backoff means that we won't get the weird throttling errors in the e2e tests anymore. 

For more on the reasoning (this PR uses a "full jitter" strategy), see [this article](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/) on the AWS architecture blog. 

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bvtujo committed:
chore(stack): validate path and root_directory against allowed characters (#1956)

This PR adds limits to the length of input paths corresponding to the limits of docker containers and EFS volumes. It also ensures that paths do not include special characters beyond a small allowlist of common characters and the `/` separator. 

<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

efekarakus committed:
chore(template): add Network opts to place tasks in private subnets  (#1968)

Add the `NetworkOpts` type to the `template` pkg so that we can pass placement information from manifests:
```go
type NetworkOpts struct {
	AssignPublicIP   string
	SubnetsType      string
	SecurityGroups []string
}
```
Also fix our integ tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bvtujo committed:
feat: enable mounting external EFS volumes in Copilot jobs & services (#1971)

<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

huanjani committed:
feat(cicd): enable Bitbucket as a source for pipelines (#1944)

Copilot uses CodeStar Connections to allow Bitbucket as a pipeline repo source. After running `pipeline update`, users update their connection status (and install the host app, if necessary), while Copilot waits. Once the updated status is detected, the pipeline is retried. 

Resolves #1274 
Related #1339

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

efekarakus committed:
chore(template): fix statemachine and sidecar so e2e tests pass (#1977)

The statemachine's `DefinitionSubstitutions` sections expects strings as values. In #1968, we updated them to be a list  and broke the template.  
We also added `"-}}"` in our service templates above the sidecar template which removed extra whitespaces resulting in the template for sidecars to be malformed, this change adds the new lines back.



By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

huanjani committed:
fix(cicd): truncate pipeline name to prevent failures (#1957)

When building pipelines, we are limited by character maximums for multiple strings:
128 for PipelineRolePolicy, BuildProjectPolicy, etc., and the CODEBUILD_BUILD_ID tag for Docker.
100 for CodePipeline pipeline names.

So that pipelines built from repos with very long names and/or repo provider accounts with very long names do not fail, this change reduces the characters when we generate pipeline names, and throws an error when pipeline names exceed 100 characters in `pipeline update` (in case pipeline names were edited in the pipeline manifest). We limit to 100 to satisfy CodePipeline's pipeline name limit (100), but also to enable the automated tacking on of `-CodepipelinePolicy` and `-BuildprojectPolicy` (and any future policy name suffixes) to the end of the stack names (128). 

The extra characters are omitted _from the front_; at the moment, pipeline names go from general ("pipeline") to specific (repoName or even branchName, if users are [changing pipeline names to enable multiple pipelines from a single repo](https://github.com/aws/copilot-cli/discussions/1925)) so truncating from the front of the string preserves the more valuable information.

In the buildspec, we truncate the string yet again, this time to 128 characters. Again, we cut off excess characters from the front. Here, it is because `-BuildProject-[UUID]` is appended to form the CODEBUILD_BUILD_ID, and we don't want to lose the identifying UUID. We must truncate in two steps because it is possible for a pipeline name (of, say, 100 characters), to successfully generate a CFN stack, but fail in CodeBuild because with the added characters it becomes 150 characters long. 

Fixes #1939.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

efekarakus committed:
feat(manifest): enable "network" field to specify subnets and sec groups (#1979)

The manifests now support the following new fields:
```yaml
network:
   vpc:
      placement: 'public|private'  # which subnets to place the tasks.
      security_groups: [string]    # list of additional security groups.
```

The `network.vpc.placement` field won't be really useful atm for customers that use Copilot generated VPCs since we don't create NAT gateways. However, it will be supported in the future. This is milestone 1. of #1959.  
On the other hand, customers that imported their VPC while creating an environment will now be able to place their tasks in private subnets.

The security groups feature is available for everyone to attach additional security groups to their tasks.

Related #1959
Closes #1453

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

Lou1415926 commented on site/content/docs/concepts/pipelines.md:
```diff
@@ -93,27 +93,31 @@ stages:
+### Step 4: Pushing New Files to your Repository
 
-Now that your `pipeline.yml` and `buildspec.yml` are created, check them in and push them to your repository. The `buildspec.yml` is needed for your pipeline's `build` stage to run successfully. Once you've done that, to actually create your pipeline run:
+Now that your `pipeline.yml`, `buildspec.yml`, and `.workspace` files have been created, add them to your repository. The entire `copilot/` directory is needed for your pipeline's `build` stage to run successfully. 
```
I am wondering If we need everything under the `copilot/` directory, should we also modify the instructions in "Creating a Pipeline in 3 steps"? 

It seems a little unclear to me whether we just need `pipeline.yml`, `buildspec.yml` and `.workspace` (since in the 3 steps instructions it only included these files), or we need the entire directory (as mentioned here)

huanjani replied:
Ah, good question! Will fix to make more specific!

Lou1415926 reviewed (commented):
LGTM!! Just a question about the instruction ⬇️

huanjani committed:
Merge remote-tracking branch 'origin/pipeline-files-messaging' into pipeline-files-messaging

huanjani committed:
chore: tweak wording to specify files

huanjani committed:
Revert "chore: tweak wording to specify files"

This reverts commit 23d0662f21b35c37bf94fb2a1ab9f680f2335b8b.

huanjani committed:
Revert " feat(manifest): enable "network" field to specify subnets and sec groups (#1979)"

This reverts commit 8bee9aa19cf24cb92a05e2117e1e8a9f64d6b750.

huanjani commented:
Sorry. Made a big git mess. Closing in favor of a new PR #1981.
